File: Factorial.java
import java.math.BigInteger;

/**
 * Factorial calculation class with both iterative and recursive implementations.
 */
public class Factorial {
    
    /**
     * Calculate factorial using recursion.
     * 
     * @param n Non-negative integer to calculate factorial of
     * @return The factorial of n (n!)
     * @throws IllegalArgumentException If n is negative
     * @throws StackOverflowError If n is too large for recursion
     */
    public static BigInteger factorialRecursive(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("Factorial is not defined for negative numbers");
        }
        
        // Base cases
        if (n == 0 || n == 1) {
            return BigInteger.ONE;
        }
        
        // Recursive case
        return BigInteger.valueOf(n).multiply(factorialRecursive(n - 1));
    }
    
    /**
     * Calculate factorial using iteration.
     * 
     * @param n Non-negative integer to calculate factorial of
     * @return The factorial of n (n!)
     * @throws IllegalArgumentException If n is negative
     */
    public static BigInteger factorialIterative(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("Factorial is not defined for negative numbers");
        }
        
        BigInteger result = BigInteger.ONE;
        for (int i = 2; i <= n; i++) {
            result = result.multiply(BigInteger.valueOf(i));
        }
        
        return result;
    }
    
    /**
     * Calculate factorial using the specified method.
     * 
     * @param n Non-negative integer to calculate factorial of
     * @param method Either "iterative" or "recursive" (default: "iterative")
     * @return The factorial of n (n!)
     * @throws IllegalArgumentException If n is negative or method is invalid
     */
    public static BigInteger factorial(int n, String method) {
        switch (method) {
            case "iterative":
                return factorialIterative(n);
            case "recursive":
                return factorialRecursive(n);
            default:
                throw new IllegalArgumentException(
                    "Invalid method: " + method + ". Use 'iterative' or 'recursive'"
                );
        }
    }
    
    /**
     * Calculate factorial using iterative method (default).
     */
    public static BigInteger factorial(int n) {
        return factorial(n, "iterative");
    }
    
    public static void main(String[] args) {
        System.out.println("Factorial Calculator Demo");
        System.out.println("========================================");
        
        // Test cases
        int[] testValues = {0, 1, 5, 10, 20};
        
        for (int val : testValues) {
            try {
                BigInteger iterResult = factorial(val, "iterative");
                BigInteger recResult = factorial(val, "recursive");
                System.out.printf("%d! = %s (iterative) = %s (recursive)%n", 
                    val, iterResult, recResult);
            } catch (Exception e) {
                System.err.printf("Error calculating %d!: %s%n", val, e.getMessage());
            }
        }
        
        // Test error handling
        System.out.println("\nError Handling Tests:");
        System.out.println("----------------------------------------");
        
        // Test negative number
        try {
            factorial(-5);
        } catch (IllegalArgumentException e) {
            System.out.println("✓ Caught error for negative input: " + e.getMessage());
        }
        
        // Test invalid method
        try {
            factorial(5, "invalid");
        } catch (IllegalArgumentException e) {
            System.out.println("✓ Caught error for invalid method: " + e.getMessage());
        }
        
        // Test recursion limit (Java handles large recursion better than Python)
        System.out.println("\nTesting large factorial...");
        try {
            BigInteger result = factorial(1000, "iterative");
            System.out.printf("1000! has %d digits%n", result.toString().length());
        } catch (Exception e) {
            System.err.println("Error with large factorial: " + e.getMessage());
        }
        
        // Very large factorial
        int largeN = 5000;
        BigInteger result = factorial(largeN, "iterative");
        System.out.printf("%d! has %d digits%n", largeN, result.toString().length());
    }
}

File: factorial.py
#!/usr/bin/env python3
"""
Factorial calculation module with both iterative and recursive implementations.
"""

from typing import Union


def factorial_recursive(n: int) -> int:
    """
    Calculate factorial using recursion.
    
    Args:
        n: Non-negative integer to calculate factorial of
        
    Returns:
        The factorial of n (n!)
        
    Raises:
        ValueError: If n is negative
        TypeError: If n is not an integer
        RecursionError: If n is too large for Python's recursion limit
    """
    if not isinstance(n, int):
        raise TypeError(f"Expected int, got {type(n).__name__}")
    
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers")
    
    # Base cases
    if n in (0, 1):
        return 1
    
    # Recursive case
    try:
        return n * factorial_recursive(n - 1)
    except RecursionError:
        raise RecursionError(f"Input {n} is too large for recursive calculation")


def factorial_iterative(n: int) -> int:
    """
    Calculate factorial using iteration.
    
    Args:
        n: Non-negative integer to calculate factorial of
        
    Returns:
        The factorial of n (n!)
        
    Raises:
        ValueError: If n is negative
        TypeError: If n is not an integer
    """
    if not isinstance(n, int):
        raise TypeError(f"Expected int, got {type(n).__name__}")
    
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers")
    
    result = 1
    for i in range(2, n + 1):
        result *= i
    
    return result


def factorial(n: int, method: str = "iterative") -> int:
    """
    Calculate factorial using the specified method.
    
    Args:
        n: Non-negative integer to calculate factorial of
        method: Either "iterative" or "recursive" (default: "iterative")
        
    Returns:
        The factorial of n (n!)
        
    Raises:
        ValueError: If n is negative or method is invalid
        TypeError: If n is not an integer
        RecursionError: If using recursive method with large n
        
    Examples:
        >>> factorial(5)
        120
        >>> factorial(5, "recursive")
        120
        >>> factorial(0)
        1
        >>> factorial(-1)
        Traceback (most recent call last):
            ...
        ValueError: Factorial is not defined for negative numbers
    """
    if method == "iterative":
        return factorial_iterative(n)
    elif method == "recursive":
        return factorial_recursive(n)
    else:
        raise ValueError(f"Invalid method: {method}. Use 'iterative' or 'recursive'")


def main():
    """Demonstrate factorial calculations."""
    print("Factorial Calculator Demo")
    print("=" * 40)
    
    # Test cases
    test_values = [0, 1, 5, 10, 20]
    
    for val in test_values:
        try:
            iter_result = factorial(val, "iterative")
            rec_result = factorial(val, "recursive")
            print(f"{val}! = {iter_result} (iterative) = {rec_result} (recursive)")
        except Exception as e:
            print(f"Error calculating {val}!: {e}")
    
    # Test error handling
    print("\nError Handling Tests:")
    print("-" * 40)
    
    # Test negative number
    try:
        factorial(-5)
    except ValueError as e:
        print(f"✓ Caught error for negative input: {e}")
    
    # Test non-integer
    try:
        factorial(5.5)
    except TypeError as e:
        print(f"✓ Caught error for float input: {e}")
    
    # Test invalid method
    try:
        factorial(5, "invalid")
    except ValueError as e:
        print(f"✓ Caught error for invalid method: {e}")
    
    # Test recursion limit
    print("\nTesting recursion limit...")
    try:
        factorial(10000, "recursive")
    except RecursionError as e:
        print(f"✓ Caught recursion error: {e}")
    
    # Large factorial with iterative method
    large_n = 100
    result = factorial(large_n, "iterative")
    print(f"\n{large_n}! has {len(str(result))} digits")


if __name__ == "__main__":
    main()